src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java

```diff
@@ -663,23 +728,44 @@ private void onSignal(
 
             switch ((int) signalId)
             {
-            case TOKEN_EXPIRED_SIGNAL:
-                onTokenExpiredSignal(signal);
+            case GRANT_VALIDATION_SIGNAL:
+                onGrantValidationSignal(signal);
                 break;
             default:
                 break;
             }
         }
 
-        private void onTokenExpiredSignal(
+        private void onGrantValidationSignal(
             SignalFW signal)
         {
-            final long delay = grant.expiresAt - System.currentTimeMillis();
+            final long now = System.currentTimeMillis();
+            long delay = grant.expiresAtMillis - now;
 
             if (delay > 0)
             {
-                this.expiryFuture = executor.schedule(delay, MILLISECONDS,
-                        targetRouteId, targetStreamId, TOKEN_EXPIRED_SIGNAL);
+                final long challengeAfter = grant.challengeAfter();
+
+                if (canChallenge(capabilities))
+                {
+                    if (grant.withinChallengeResponseTimeout(now))
+                    {
+                        // Check if a challenge hasn't been sent yet. If not, update challenged at and send the challenge
+                        if (grant.lastChallengedAt < challengeAfter)
+                        {
+                            grant.lastChallengedAt = now;
+                            doChallenge(signal.trace());
```

Comment:
Realize `doChallenge` isn't visible to shared `OAuthAccessGrant`, but couldn't we pass in `long traceId` and a `LongConsumer challenger` to `grant.challenge(now, traceId, this::doChallenge)`, then move this logic to `challenge` method to update internal `lastChallengedAt` state.

Then `challenge` method can fold in `withinChallengeResponseTimeout` conditional logic as well.

The return value can be the `nextSignalDelay` used for the `signalFuture`.